documentation/projects/proposals/additional_search_views/20230719-implementation_plan_additional_search_views.md

```diff
@@ -0,0 +1,366 @@
+# 2023-07-20 Implementation Plan: Additional Search Views
+
+**Author**: @obulat
+
+<!-- See the implementation plan guide for more information: https://github.com/WordPress/openverse/tree/19791f51c063d0979112f4b9f4eeace04c8cf5ff/docs/projects#implementation-plans-status-in-rfc -->
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @zackkrida
+- [ ] @sarayourfriend
+
+## Project links
+
+<!-- Enumerate any references to other documents/pages, including milestones and other plans -->
+
+- [Project Thread](https://github.com/WordPress/openverse/issues/410)
+- [Project Proposal](https://docs.openverse.org/projects/proposals/additional_search_views/20230424-project_proposal_additional_search_views.html)
+
+## Overview
+
+<!-- An overview of the implementation plan, if necessary. Save any specific steps for the section(s) below. -->
+
+This plan describes the changes that need to be added on the frontend: the new
+components, pages and store.
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+Three collection pages are added to Openverse:
+
+- items with the selected tag
+- items from the selected source (provider)
+- items by the selected creator.
+
+The pages re-use the existing image grid and audio collection views and can load
+more images or audio items on the Load more button click. The single result
+pages are updated to add the links to these pages, as seen in the Figma mockups:
+
+- [Image single result](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-63284&mode=dev)
+- [Audio single result](https://www.figma.com/file/niWnCgB7K0Y4e4mgxMrnRC/Additional-search-views?type=design&node-id=1200-63285&mode=dev)
+
+For the creator pages, a separate API media view is added to get the items that
+match the creator and source pair exactly.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+### API changes
```

Comment:
Thank you for adding this section. It helps a lot to understand how this will all fit together.

I like the approach to add new routes on the API rather than trying to stuff everything into the existing search endpoint. That makes things a lot nicer for the creator and provider especially.

However, there are some problems with using the database rather than ES to access anything:

1. The database does not cache queries in the same way that ES does. Repeated queries will not necessarily be as efficient as from ES.
2. The database does not score documents at all, so the order will different dramatically to the way that ES would order the documents. That's an issue with respect to popularity data today already, but will become even more of an issue if we start to score documents based on other metrics as theorised by our search relevancy discussions.
3. `creator` is not indexed in the API database, so a query against it will be very slow.

Only one of these can be "fixed", the third, but it's much easier to add a `creator.raw` field to the index instead, like we've done for title, description, and tag name. Adding a new field to the index is far easier than adding a new index to the database (which would require careful application to avoid a long-running migration) and which could significantly increase the time it takes to do a data refresh.

I'd also strongly recommend against the route described as it is harder to cache in a sensible way (on the Cloudflare side of things) or to perform cache invalidation that might be required by #1969. Instead of using query strings, we can describe the resource via the path: `/<media type>/source/<source>/creator/<creator>` is very clean, easy to read and understand, and very easy to manage the cache for because it is a static path. The source page can use the same route by leaving off the creator. This removes the need to manage specific query params as well and would allow us to add querying within these routes more easily in the future behind the regular `q` parameter if we wanted.

For the tag route, I'd recommend the singular `tag` rather than plural `tags`, for legibility if we are presenting a single tag. Again, using a route-based approach instead of query parameter approach makes it easier to manage the cache if needed and could allow us to more easily feature specific tags with high quality results as example pages if we wanted, like `/images/tag/dog` or `/images/tag/flying%20birds`.

Path elements still need to be URL encoded to preserve special characters and spaces, but for most cases like `/images/source/nasa`, `/images/sources/nypl`, `/images/sources/inaturalist`, etc we've suddenly got a very good-looking URL to point people to.